observatory_platform/telescopes/crossref_events.py

```diff
@@ -0,0 +1,710 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Author: Aniek Roelofs
+
+import fileinput
+import json
+import logging
+import os
+import pathlib
+import shutil
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from datetime import timedelta
+from typing import Tuple, Union
+
+import pendulum
+from airflow.exceptions import AirflowException
+from airflow.hooks.base_hook import BaseHook
+from airflow.models import Variable
+from airflow.models.taskinstance import TaskInstance
+from google.cloud import bigquery
+from google.cloud.bigquery import SourceFormat
+from requests.exceptions import RetryError
+
+from observatory_platform.utils.config_utils import (AirflowConn,
+                                                     AirflowVar,
+                                                     SubFolder,
+                                                     check_connections,
+                                                     check_variables,
+                                                     find_schema,
+                                                     schema_path,
+                                                     telescope_path,
+                                                     telescope_templates_path)
+from observatory_platform.utils.gc_utils import (bigquery_partitioned_table_id,
+                                                 create_bigquery_dataset,
+                                                 load_bigquery_table,
+                                                 run_bigquery_query,
+                                                 upload_file_to_cloud_storage)
+from observatory_platform.utils.jinja2_utils import (make_sql_jinja2_filename, render_template)
+from observatory_platform.utils.url_utils import retry_session
+
+
+def extract_events(url: str, events_path: str, next_cursor: str = None, success: bool = True) -> \
+        Tuple[bool, Union[str, None], Union[int, None]]:
+    """
+    Extract the events from the given url until no new cursor is returned or a RetryError occurs. The extracted events
+    are appended to a json file, with 1 list per request.
+
+    :param url: The crossref events api url
+    :param events_path: Path to the file in which events are stored
+    :param next_cursor: The next cursor, this is in the response of the api
+    :param success: Whether all events were extracted successfully or an error occurred
+    :return: success, next_cursor and number of total events
+    """
+    headers = {
+        'User-Agent': 'Observatory Platform (+https://github.com/The-Academic-Observatory/observatory-platform; '
+                      f'mailto:{url.split("mailto=")[1].split("&")[0]})'
+    }
+    tmp_url = url + f'&cursor={next_cursor}' if next_cursor else url
```

Comment:
Tuan made a function `get_ao_user_agent` that has been merged with develop: https://github.com/The-Academic-Observatory/observatory-platform/blob/b62d87c07531f7ba35272393dca6da7734cf5f41/observatory_platform/utils/url_utils.py#L92

It might be a good idea to use and or extend this function to build the user agent string for standardisation purposes. You could add a keyword argument for the email address if you want to be able customise the email address (if the keyword argument email address is None then use the pkg_info email address).

Reply:
Thanks, I'll definitely add this.